src/Corporation/ui/modals/BuybackSharesModal.tsx

```diff
@@ -23,70 +23,85 @@ export function BuybackSharesModal(props: IProps): React.ReactElement {
   const corp = useCorporation();
   const [shares, setShares] = useState<number>(NaN);
 
-  const currentStockPrice = corp.sharePrice;
-  const buybackPrice = currentStockPrice * 1.1;
-  const disabledText = !isPositiveInteger(shares)
-    ? "Number of shares must be a positive integer"
-    : shares > corp.issuedShares
-    ? "There are not enough shares available to buyback this many"
-    : shares * buybackPrice > Player.money
-    ? "Insufficient player funds"
-    : "";
+  const [cost, sharePrice] = corp.calculateShareBuyback((props.open && shares) || 0);
+  const disabledText = buybackSharesFailureReason(corp, shares);
 
   function buy(): void {
-    if (disabledText) return;
     try {
       BuyBackShares(corp, shares);
+      dialogBoxCreate(
+        <>
+          <Typography>
+            You bought {formatShares(shares)} shares for <Money money={cost} />.
+          </Typography>
+          <Typography>
+            <b>{corp.name}</b>'s stock price rose to <Money money={sharePrice} /> per share.
+          </Typography>
+        </>,
+      );
+      props.onClose();
+      props.rerender();
+      setShares(NaN);
     } catch (err) {
-      dialogBoxCreate(err + "");
+      dialogBoxCreate(`${err}`);
     }
-    props.onClose();
-    props.rerender();
   }
 
-  function CostIndicator(): React.ReactElement {
-    if (shares === null) return <></>;
-    if (isNaN(shares) || shares <= 0) {
-      return <>ERROR: Invalid value entered for number of shares to buyback</>;
-    } else if (shares > corp.issuedShares) {
-      return (
-        <>
-          There are not this many shares available to buy back. There are only {formatBigNumber(corp.issuedShares)}{" "}
-          outstanding shares.
-        </>
-      );
+  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
+    if (event.key === KEY.ENTER) buy();
+  }
+
+  interface IEffectTextProps {
+    shares: number;
+  }
+
+  function EffectText({ shares }: IEffectTextProps): React.ReactElement {
+    if (!shares) {
+      return <Typography>&nbsp;</Typography>;
+    } else if (disabledText) {
+      return <Typography>{disabledText}</Typography>;
     } else {
       return (
-        <>
-          Purchase {shares} shares for a total of {formatMoney(shares * buybackPrice)}
-        </>
+        <Typography>
+          <b>{corp.name}</b>'s stock price will rise to <Money money={sharePrice} /> per share.
+        </Typography>
       );
     }
   }
```

Comment:
A react component should never be defined inside of another react component, they should only be defined at top level scope within a module.

But I'm also not sure why this needs to be a separate component. I believe it would function identically if you just rendered that content directly within BuyBackSharesModal.

Same comment applies for other modals where this same sort of thing was implemented.